include/dca/phys/dca_step/cluster_solver/shared_tools/accumulation/tp/tp_accumulator_gpu.hpp

```diff
@@ -450,15 +493,90 @@ void TpAccumulator<Parameters, linalg::GPU>::finalize() {
     return;
 
   for (std::size_t channel = 0; channel < G4_.size(); ++channel)
+  {
+    if(distributed_g4_enabled_)
+    {
+      // modify G4 size in G4 cpu, otherwise, copyTo() operation failed due to incomparable size
+      // resize() only modifies member Nb_elements in function, does not change tp_dmn.get_size()
+      G4_[channel].resize(get_G4()[channel].size());
+    }
     get_G4()[channel].copyTo(G4_[channel]);
-
+  }
   // TODO: release memory if needed by the rest of the DCA loop.
   // get_G4().clear();
 
   finalized_ = true;
   initialized_ = false;
 }
 
+template <class Parameters>
+void TpAccumulator<Parameters, linalg::GPU>::ringG(float& flop) {
+
+//    std::array<std::pair<int, int>, 2> G2_sz;
+
+    // get ready for send and receive
+    for (int s = 0; s < 2; ++s)
+    {
+        sendbuff_G_[s] = G_[s];
+    }
+
+    int my_concurrency_id, mpi_size;
+    MPI_Comm_size(MPI_COMM_WORLD, &mpi_size);
+    MPI_Comm_rank(MPI_COMM_WORLD, &my_concurrency_id);
+
+    MPI_Request recv_request_1, recv_request_2;
+    MPI_Request send_request_1, send_request_2;
+    MPI_Status status_1, status_2, status_3, status_4;
+
+    // get rank index of left and right neighbor
+    auto mod_op = [](int id, int mpi_size) {return id % mpi_size; };
+    int left_neighbor = mod_op((my_concurrency_id - 1 + mpi_size), mpi_size);
+    int right_neighbor = mod_op((my_concurrency_id + 1 + mpi_size), mpi_size);
+
+    // Pipepline ring algorithm in the following for-loop:
+    // 1) At each time step, local rank receives a new G2 from left hand neighbor,
+    // makes a copy locally and uses this G2 to update G4, and
+    // sends this G2 to right hand neighbor. In total, the algorithm performs (mpi_size - 1) steps.
+    // 2) This algorithm currently requires parameters in input file, please refer to mci_parameters.hpp:
+    //      a) #walker == #accumulator and shared-walk-and-accumulation-thread = true;
+    //      b) and, local measurements are equal, and each accumulator should have same #measurement, i.e.
+    //         measurements % ranks == 0 && local_measurement % threads == 0.
+    for(int icount=0; icount < (mpi_size-1); icount++)
+    {
+        MPI_Irecv(G_[0].ptr(), (G_[0].size().first)*(G_[0].size().second),
```

Comment:
all of this should be hidden in a concurrency subclass. The scalar type of G and G4 can either be complex<double> or complex<float>